### ISSUE TITLE
Pinning major version

### ISSUE THREAD

**Issue by ewels** [nextflow-io/plugins#20]
## Problem

When adding a plugin to a pipeline we have to pin a specific version. This is good for reproducibility, but bad for bug fixing. We're about to add `nf-validation` into the nf-core template, but we know very well that we will hit lots of bugs and need to do patch releases. This will require a lot of manual work updating the template and many pipelines on a regular basis.

### Language
A classic way is to give syntax to say "this version or above" (eg. Nextlow itself handles this with `>=` and `!>=`). This would presumably require a language change in Nextflow to allow inclusion of pipelines like this.

### Major version pins

Possibly a simpler way is to use major version pinning. This is done with GitHub Actions (see official [GitHub docs](https://docs.github.com/en/actions/creating-actions/about-custom-actions#using-tags-for-release-management)): when a new release is made, a new specific git tag is added (`v1.4.2`) but also the tag for the major version is bumped to point to the new release (`v1`). That way, users can pin either the specific release, or they can pin `v1` and get updated patch releases as they are available, automatically.

A similar approach should be viable for Nextflow plugins I think - add a `v1` major version to the JSON file and update the download link and sha every time a new release is made. Then include the plugin in the pipeline with the major version tag.

Any thoughts @pditommaso @bentsherman? Tagging @mirpedrol @mashehu for visibility.

**Comment by pditommaso** [comment]
Default plugins are pinned in the nextflow [runtime](https://github.com/nextflow-io/nextflow/blob/ba9b6d180f3a17429a7a8c9c7c9decf8e25cfe60/modules/nextflow/src/main/resources/META-INF/plugins-info.txt#L1-L8). 

When using non-default plugins, nextflow pulls the latest version unless you specify a specific on in the config file.

**Comment by ewels** [comment]
Ah ok, I didn't realise it worked at all if you didn't pin anything. That's a good start 👍🏻 

What do you think about allowing the JSON to include a major version tag that corresponds to the latest minor / patch release for that major version?

**Comment by pditommaso** [comment]
The semantic version of each release should already be doing that, no?

**Comment by pditommaso** [comment]
A related problem is the plugin version also depends on the nextflow runtime, I mean, plugin X can require a nextflow version Y as minimal version. 

Pinning the plugin can break changing the nextflow version, therefore it would be better to avoid pinning at all.

Nextflow pick the latest version fulfilling the declared nextflow requirement e.g. 

https://github.com/nextflow-io/plugins/blob/main/plugins.json#L353

**Comment by ewels** [comment]
Right, but not pinning the plugin can break the pipeline due to breaking changing the plugin. For example, changes in how the nf-validation `fromSamplesheet()` channel factory syntax works. So you're screwed either way 😅 

That's what this approach is trying to solve ☝🏻  Not pinning the plugin, apart from major (breaking) changes in the plugin.

**Comment by pditommaso** [comment]
Moving to nextflow repo
